src/vm/gcenv.ee.cpp

```diff
@@ -830,53 +830,84 @@ void GCToEEInterface::StompWriteBarrier(WriteBarrierParameters* args)
         }
 #endif // FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
 
-        stompWBCompleteActions |= ::StompWriteBarrierResize(args->is_runtime_suspended, args->requires_upper_bounds_check);
-
-        // We need to make sure that other threads executing checked write barriers
-        // will see the g_card_table update before g_lowest/highest_address updates.
-        // Otherwise, the checked write barrier may AV accessing the old card table
-        // with address that it does not cover.
-        //
-        // Even x86's total store ordering is insufficient here because threads reading
-        // g_card_table do so via the instruction cache, whereas g_lowest/highest_address
-        // are read via the data cache.
-        //
-        // The g_card_table update is covered by section 8.1.3 of the Intel Software
-        // Development Manual, Volume 3A (System Programming Guide, Part 1), about
-        // "cross-modifying code": We need all _executing_ threads to invalidate
-        // their instruction cache, which FlushProcessWriteBuffers achieves by sending
-        // an IPI (inter-process interrupt).
+        stompWBCompleteActions |= ::StompWriteBarrierResize(is_runtime_suspended, args->requires_upper_bounds_check);
+        is_runtime_suspended = (stompWBCompleteActions & SWB_EE_RESTART) || is_runtime_suspended;
 
         if (stompWBCompleteActions & SWB_ICACHE_FLUSH)
         {
-            // flushing icache on current processor (thread)
+            // flushing/invalidating the write barrier's body for the current process 
+            // NOTE: the underlying API may flush more than needed or nothing at all if Icache is coherent.
             ::FlushWriteBarrierInstructionCache();
-            // asking other processors (threads) to invalidate their icache
+        }
+
+        // NB: managed heap segments may surround unmanaged/stack segments. In such cases adding another managed heap segment
```

Comment:
Nit: What is your rule of when to use `NB` ? I am not sure whether it is useful - but I am may be missing the purpose.

Reply:
I treat it as a stronger form of  `NOTE:` - a nonobvious fact that could lead to bugs if not considered. 
Maybe there is a better convention?

---
In reply to: [295930082](https://github.com/dotnet/coreclr/pull/25130#discussion_r295930082) [](ancestors = 295930082)

Reply:
I see `IMPORTANT:`   and block-style   `// **** IMPORTANT NOTE ****` used in other places. 

I will make this `IMPORTANT:`

---
In reply to: [295932830](https://github.com/dotnet/coreclr/pull/25130#discussion_r295932830) [](ancestors = 295932830,295930082)